File: piper/android/app/src/main/java/io/gopiper/piper/engine/Piper.kt
package io.gopiper.piper.engine

import android.app.AlertDialog
import android.app.DownloadManager
import android.content.Context
import android.net.Uri
import android.os.Environment
import android.util.Log
import android.webkit.CookieManager
import android.webkit.URLUtil
import android.webkit.WebStorage
import kotlinx.coroutines.*
import io.gopiper.piper.cheese.Browser
import io.gopiper.piper.cheese.js
import io.gopiper.piper.engine.model.Func
import io.gopiper.piper.engine.model.HTMLElement
import io.gopiper.piper.engine.statements.*
import io.gopiper.piper.model.Pipe
import io.gopiper.piper.util.C
import io.pipend.open.hiper.Hiper
import me.gumify.hiper.util.WeeDB
import me.gumify.hiper.util.onUiThread
import me.gumify.hiper.util.toast
import org.json.JSONArray
import org.json.JSONObject
import java.lang.Exception


// 871 6340 3444
// 245256

class Piper(
    val context: Context,
    val browser: Browser,
    val onScriptComplete: () -> Unit,
    val onRegisterLog: (scriptId: String, log: String) -> Unit,
    val isHeadless: Boolean = false,
    val scriptId: String
) {
    companion object {
        const val VERSION = 1
    }

    val hiper = Hiper.getInstance()
    val wee = WeeDB(context.applicationContext)


    var loopFlow: String? = null
    var currentFunctionName: String? = null
    var isFunctionReturn: Boolean? = null
    var functionReturnValue: Any? = null

    val VARIABLE_STACK = HashMap<String, Any?>()
    val FUNCTION_STACK = ArrayList<Func>()
    var isTerminated = false
    private val isDoNotWait = wee.getBoolean(C.SETTING_DO_NOT_WAIT_FOR_LOADING)
    private val isSkipError = wee.getBoolean(C.SETTING_SKIP_ERROR)
    var canShowErrorAlert = true
    var lastClickedElement: HTMLElement? = null

    fun loadUrl(url: String) {
        browser.currentTab.value?.loadUrl(url)
    }

    suspend fun downloadFromUrl(url: String) {
        var isDone = false
        browser.currentTab.value?.webView?.js("""
            var el = document.createElement("a");
            el.href = "$url";
            el.setAttribute("download", true);
            el.click();
        """.trimIndent()) {
            isDone = true
        }
        while (!isDone) {
            if (isTerminated) break
            delay(100)
        }

//        val response = hiper.head(url)
//        val fileName = URLUtil.guessFileName(url, response.headers["content-disposition"], response.headers["content-type"])
//        val downloadmanager = context.getSystemService(Context.DOWNLOAD_SERVICE) as? DownloadManager
//        val uri: Uri = Uri.parse(url)
//        onUiThread {
//            val request = DownloadManager.Request(uri)
//            request.setMimeType(response.headers["content-type"])
//            val cookies = CookieManager.getInstance().getCookie(url)
//            request.addRequestHeader("cookie", cookies)
//            request.addRequestHeader("User-Agent", browser.currentTab.value?.webView?.settings?.userAgentString)
//            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
//            request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, fileName)
//            downloadmanager?.enqueue(request)
//        }
    }

    suspend fun findFunction(name: String?): Func? {
        if (name == null) return null
        for (f in FUNCTION_STACK) {
            if (f.name == name) return f
        }
        return null
    }

    fun showError(statement: JSONObject, error: Exception) {
        if (!isSkipError) {
            isTerminated = true
            if (!isHeadless && canShowErrorAlert) {
                onUiThread {
                    val dialog = AlertDialog.Builder(context)
                    dialog.setTitle("Error")
                    dialog.setMessage(
                        """
                ${statement.getString("cmd").uppercase()}
                
                Error: $error
            """.trimIndent()
                    )
                    dialog.setPositiveButton("OK") { _, _ -> }
                    dialog.show()
                }
            }
        }
        val log = "${statement["cmd"].toString().uppercase()}\n$error"
        onRegisterLog(scriptId, log)
    }

    suspend fun init(code: String) {
        if (wee.getBoolean(C.SETTING_CLEAR_BROWSER_BEFORE_EX)) {
            WebStorage.getInstance().deleteAllData()
            CookieManager.getInstance().removeAllCookies(null)
            CookieManager.getInstance().flush()
        }
        val isSuccess = initFunctions(code)
        delay(1000)
        if (isSuccess) {
            execute(code)
        }
        onScriptComplete()
    }

    private suspend fun initFunctions(code: String): Boolean {
        val statements: JSONArray
        try {
            statements = JSONArray(code)
        } catch (e: Exception) {
            Log.d("hello", "Error: $e")
            onUiThread {
                val o = JSONObject()
                o.put("cmd", "INIT_FUNCTIONS")
                showError(o, Exception("SyntaxError: There might be some invalid blocks"))
            }
            return false
        }
        for (i in 0 until statements.length()) {
            val statement = statements.getJSONObject(i)
            val cmd = statement.getString("cmd")
            if (cmd == "procedures_defnoreturn") {
                executeProceduresDefnoreturn(statement) { showError(statement, it) }
            } else if (cmd == "procedures_defreturn") {
                executeProceduresDefreturn(statement) { showError(statement, it) }
            }
        }
        return true
    }

    suspend fun execute(sourceCode: String) {
        val statements = JSONArray(sourceCode)
        for (i in 0 until statements.length()) {
            if (isTerminated) break

            if (loopFlow == "BREAK") {
                break
            } else if (loopFlow == "CONTINUE") {
                continue
            }

            if (isFunctionReturn != null) break

            val statement = statements.getJSONObject(i)
            executeStatement(statement)
        }
    }

    suspend fun executeStatement(statement: JSONObject): Any? {
        if (!isDoNotWait) {
            while (browser.currentTab.value?.isLoading == true) delay(100)
        }
        if (!statement.has("cmd")) {
            return statement
        }

        return when (statement.getString("cmd")) {
            // Browser
            "go_to_website" -> {
                lastClickedElement = null
                val ret = executeGoToWebsite(statement) { showError(statement, it) }
                if (!isDoNotWait) {
                    delay(1000)
                }
                return ret
            }
            "download_from_url" -> executeDownloadFromUrl(statement) { showError(statement, it) }
            "set_user_agent" -> executeSetUserAgent(statement) { showError(statement, it) }
            "custom_user_agent" -> executeCustomUserAgent(statement) { showError(statement, it) }
            "execute_javascript" -> executeExecuteJavascript(statement) { showError(statement, it) }
            "new_tab" -> executeNewTab(statement) { showError(statement, it) }
            "switch_tab" -> executeSwitchTab(statement) { showError(statement, it) }
            "close_tab" -> executeCloseTab(statement) { showError(statement, it) }

            // HTML
            "html_elements" -> executeHtmlElements(statement) { showError(statement, it) }
            "html_elements_by_selector" -> executeHtmlElementsBySelector(statement) { showError(statement, it) }
            "wait_for_elements" -> executeWaitForElements(statement) { showError(statement, it) }
            "attribute_of" -> executeAttributeOf(statement) { showError(statement, it) }
            "text_of" -> executeTextOf(statement) { showError(statement, it) }
            "tag_name_of" -> executeTagNameOf(statement) { showError(statement, it) }
            "is_html_element_exist" -> executeIsHtmlElementExist(statement) { showError(statement, it) }
            "page_source" -> executePageSource(statement) { showError(statement, it) }

            // Events
            "click" -> executeClick(statement) { showError(statement, it) }
            "type" -> executeType(statement) { showError(statement, it) }
            "press_key" -> executePressKey(statement) { showError(statement, it) }
            "wait_for" -> executeWaitFor(statement) { showError(statement, it) }
            "scroll" -> executeScroll(statement) { showError(statement, it) }

            // Logic
            "controls_if" -> executeControlsIf(statement) { showError(statement, it) }
            "logic_compare" -> executeLogicCompare(statement) { showError(statement, it) }
            "logic_operation" -> executeLogicOperation(statement) { showError(statement, it) }
            "logic_negate" -> executeLogicNegate(statement) { showError(statement, it) }
            "logic_boolean" -> executeLogicBoolean(statement) { showError(statement, it) }
            "logic_null" -> null
            "logic_ternary" -> executeLogicTernary(statement) { showError(statement, it) }

            // Loops
            "controls_repeat_ext" -> executeControlsRepeatExt(statement) { showError(statement, it) }
            "controls_whileUntil" -> executeControlsWhileUntil(statement) { showError(statement, it) }
            "controls_for" -> executeControlsFor(statement) { showError(statement, it) }
            "controls_forEach" -> executeControlsForEach(statement) { showError(statement, it) }
            "controls_flow_statements" -> executeControlsFlowStatement(statement) { showError(statement, it) }

            // Math
            "math_number" -> executeMathNumber(statement) { showError(statement, it) }
            "math_arithmetic" -> executeMathArithmetic(statement) { showError(statement, it) }
            "math_single" -> executeMathSingle(statement) { showError(statement, it) }
            "math_trig" -> executeMathTrig(statement) { showError(statement, it) }
            "math_constant" -> executeMathConstant(statement) { showError(statement, it) }
            "math_number_property" -> executeMathNumberProperty(statement) { showError(statement, it) }
            "convert_text_to_number" -> executeConvertTextToNumber(statement) { showError(statement, it) }
            "math_round" -> executeMathRound(statement) { showError(statement, it) }
            "math_on_list" -> executeMathOnList(statement) { showError(statement, it) }
            "math_modulo" -> executeMathModulo(statement) { showError(statement, it) }
            "math_constrain" -> executeMathConstrain(statement) { showError(statement, it) }
            "math_random_int" -> executeMathRandomInt(statement) { showError(statement, it) }
            "math_random_float" -> executeMathRandomFloat(statement) { showError(statement, it) }

            // Text
            "text" -> executeText(statement) { showError(statement, it) }
            "text_join" -> executeTextJoin(statement) { showError(statement, it) }
            "text_append" -> executeTextAppend(statement) { showError(statement, it) }
            "text_length" -> executeTextLength(statement) { showError(statement, it) }
            "text_isEmpty" -> executeTextIsEmpty(statement) { showError(statement, it) }
            "convert_to_text" -> executeConvertToText(statement) { showError(statement, it) }
            "text_indexOf" -> executeTextIndexOf(statement) { showError(statement, it) }
            "text_charAt" -> executeTextCharAt(statement) { showError(statement, it) }
            "text_getSubstring" -> executeTextGetSubstring(statement) { showError(statement, it) }
            "text_changeCase" -> executeTextChangeCase(statement) { showError(statement, it) }
            "text_trim" -> executeTextTrim(statement) { showError(statement, it) }
            "text_print" -> executeTextPrint(statement) { showError(statement, it) }
            "prompt" -> executePrompt(statement) { showError(statement, it) }

            // Lists
            "lists_create_with" -> executeListsCreateWith(statement) { showError(statement, it) }
            "lists_chooser" -> executeListsChooser(statement) { showError(statement, it) }
            "lists_repeat" -> executeListsRepeat(statement) { showError(statement, it) }
            "lists_length" -> executeListsLength(statement) { showError(statement, it) }
            "lists_isEmpty" -> executeListsIsEmpty(statement) { showError(statement, it) }
            "lists_indexOf" -> executeListsIndexOf(statement) { showError(statement, it) }
            "lists_getIndex" -> executeListsGetIndex(statement) { showError(statement, it) }
            "lists_setIndex" -> executeListsSetIndex(statement) { showError(statement, it) }
            "lists_getSublist" -> executeListsGetSublist(statement) { showError(statement, it) }
            "lists_split" -> executeListsSplit(statement) { showError(statement, it) }
            "lists_regex" -> executeListsRegex(statement) { showError(statement, it) }
            "lists_sort" -> executeListsSort(statement) { showError(statement, it) }

            // Variables
            "variables_set" -> executeVariablesSet(statement) { showError(statement, it) }
            "variables_get" -> executeVariablesGet(statement) { showError(statement, it) }
            "math_change" -> executeMathChange(statement) { showError(statement, it) }

            // Functions
//            "procedures_defnoreturn" -> executeProceduresDefnoreturn(statement) { showError(statement, it) }
            "procedures_callnoreturn" -> executeProceduresCallnoreturn(statement) { showError(statement, it) }
            "procedures_ifreturn" -> executeProceduresIfreturn(statement) { showError(statement, it) }
//            "procedures_defreturn" -> executeProceduresDefreturn(statement) { showError(statement, it) }
            "procedures_callreturn" -> executeProceduresCallreturn(statement) { showError(statement, it) }

            // Utils
            "utils_tryCatch" -> executeUtilsTryCatch(statement) { showError(statement, it) }

            else -> null
        }
    }

    suspend fun executeStatement(statement: String): Any? {
        val obj: JSONObject
        try {
            obj = JSONObject(statement)
        } catch (e: Exception) {
            return statement
        }
        return executeStatement(obj)
    }
}

File: piper/extensions/chrome/src/engine/piper.js
import go_to_website from "./statements/go_to_website"
import download_from_url from "./statements/download_from_url"
import set_user_agent from "./statements/set_user_agent"
import execute_javascript from "./statements/execute_javascript"
import new_tab from "./statements/new_tab"
import switch_tab from "./statements/switch_tab"
import close_tab from "./statements/close_tab"

import html_elements from "./statements/html_elements"
import html_elements_by_selector from "./statements/html_elements_by_selector"
import wait_for_elements from "./statements/wait_for_elements"
import attribute_of from "./statements/attribute_of"
import text_of from "./statements/text_of"
import tag_name_of from "./statements/tag_name_of"
import is_html_element_exist from "./statements/is_html_element_exist"
import page_source from "./statements/page_source"

import click from "./statements/click"
import event_type from "./statements/event_type"
import press_key from "./statements/press_key"
import wait_for from "./statements/wait_for"
import scroll from "./statements/scroll"

import controls_if from "./statements/controls_if"
import logic_compare from "./statements/logic_compare"
import logic_operation from "./statements/logic_operation"
import logic_negate from "./statements/logic_negate"
import logic_boolean from "./statements/logic_boolean"
import logic_ternary from "./statements/logic_ternary"

import controls_repeat_ext from "./statements/controls_repeat_ext"
import controls_whileUntil from "./statements/controls_whileUntil"
import controls_for from "./statements/controls_for"
import controls_forEach from "./statements/controls_forEach"
import controls_flow_statements from "./statements/controls_flow_statements"

import math_number from "./statements/math_number"
import math_arithmetic from "./statements/math_arithmetic"
import math_single from "./statements/math_single"
import math_trig from "./statements/math_trig"
import math_constant from "./statements/math_constant"
import math_number_property from "./statements/math_number_property"
import convert_text_to_number from "./statements/convert_text_to_number"
import math_round from "./statements/math_round"
import math_on_list from "./statements/math_on_list"
import math_modulo from "./statements/math_modulo"
import math_constrain from "./statements/math_constrain"
import math_random_int from "./statements/math_random_int"

import text from "./statements/text"
import text_join from "./statements/text_join"
import text_append from "./statements/text_append"
import text_length from "./statements/text_length"
import text_isEmpty from "./statements/text_isEmpty"
import convert_to_text from "./statements/convert_to_text"
import text_indexOf from "./statements/text_indexOf"
import text_charAt from "./statements/text_charAt"
import text_getSubstring from "./statements/text_getSubstring"
import text_changeCase from "./statements/text_changeCase"
import text_trim from "./statements/text_trim"
import text_print from "./statements/text_print"
import prompt from "./statements/prompt"

import lists_create_with from "./statements/lists_create_with"
import lists_chooser from "./statements/lists_chooser"
import lists_repeat from "./statements/lists_repeat"
import lists_length from "./statements/lists_length"
import lists_isEmpty from "./statements/lists_isEmpty"
import lists_indexOf from "./statements/lists_indexOf"
import lists_getIndex from "./statements/lists_getIndex"
import lists_setIndex from "./statements/lists_setIndex"
import lists_getSublist from "./statements/lists_getSublist"
import lists_split from "./statements/lists_split"
import lists_regex from "./statements/lists_regex"
import lists_sort from "./statements/lists_sort"

import variables_set from "./statements/variables_set"
import variables_get from "./statements/variables_get"
import math_change from "./statements/math_change"

import procedures_defnoreturn from "./statements/procedures_defnoreturn"
import procedures_defreturn from "./statements/procedures_defreturn"


import { sleep } from "./util"


class Piper {
    constructor(browser) {
        this.browser = browser
        this.UA = null

        this.loopFlow = null
        this.currentFunctionName = null
        this.isFunctionReturn = null
        this.functionReturnValue = null

        this.VARIABLE_STACK = {}
        this.FUNCTION_STACK = []
        this.isTerminated = false
        this.canShowErrorAlert = true

        this.browser.onClose(windowId => {
            this.isTerminated = true
        })
    }

    async showError(statement, error) {
        this.isTerminated = true
    }

    async onScriptComplete() {
        console.log("Completed")
    }

    async init(code) {
        await sleep(500)
        this.browser.loadUrl("file:///C:/Piper/index.html")
        await sleep(500)
        this.browser.loadUrl("file:///C:/Piper/index.html")
        let isSuccess = this.initFunctions(code)
        if (isSuccess) {
            await this.execute(code)
        }
        await this.onScriptComplete()
    }

    async initFunctions(code) {
        let statements = null
        try {
            statements = JSON.parse(code)
        } catch (e) {
            this.showError({ cmd: "function_init" }, e)
            return false
        }

        for (let statement of statements) {
            if (statement.cmd === "procedures_defnoreturn") {
                await procedures_defnoreturn.call(this, statement, e => { this.showError(statement, e) })
            } else if (statement.cmd === "procedures_defreturn") {
                await procedures_defreturn.call(this, statement, e => { this.showError(statement, e) })
            }
        }
        return true
    }

    async execute(sourceCode) {
        let statements = JSON.parse(sourceCode)
        for (let statement of statements) {
            if (this.isTerminated) break
            if (this.loopFlow === "BREAK") {
                break
            } else if (this.loopFlow === "CONTINUE") {
                continue
            }

            if (this.isFunctionReturn !== null) break

            await this.executeStatement(statement)
        }
    }

    async executeStatement(statement) {
        await this.browser.loadingWait()
        if (typeof statement === "string") {
            try {
                statement = JSON.parse(statement)
            } catch (e) {
                return statement
            }
        }

        if (statement.cmd === undefined) {
            return statement
        }

        console.log(statement)

        switch (statement.cmd) {

            // Browser
            case "go_to_website":
                return await go_to_website.call(this, statement, e => { this.showError(statement, e) })
            case "download_from_url":
                return await download_from_url.call(this, statement, e => { this.showError(statement, e) })
            case "set_user_agent":
                return await set_user_agent.call(this, statement, e => { this.showError(statement, e) })
            case "execute_javascript":
                return await execute_javascript.call(this, statement, e => { this.showError(statement, e) })
            case "new_tab":
                return await new_tab.call(this, statement, e => { this.showError(statement, e) })
            case "switch_tab":
                return await switch_tab.call(this, statement, e => { this.showError(statement, e) })
            case "close_tab":
                return await close_tab.call(this, statement, e => { this.showError(statement, e) })


            // HTML
            case "html_elements":
                return await html_elements.call(this, statement, e => { this.showError(statement, e) })
            case "html_elements_by_selector":
                return await html_elements_by_selector.call(this, statement, e => { this.showError(statement, e) })
            case "wait_for_elements":
                return await wait_for_elements.call(this, statement, e => { this.showError(statement, e) })
            case "attribute_of":
                return await attribute_of.call(this, statement, e => { this.showError(statement, e) })
            case "text_of":
                return await text_of.call(this, statement, e => { this.showError(statement, e) })
            case "tag_name_of":
                return await tag_name_of.call(this, statement, e => { this.showError(statement, e) })
            case "is_html_element_exist":
                return await is_html_element_exist.call(this, statement, e => { this.showError(statement, e) })
            case "page_source":
                return await page_source.call(this, statement, e => { this.showError(statement, e) })


            // Events
            case "click":
                return await click.call(this, statement, e => { this.showError(statement, e) })
            case "type":
                return await event_type.call(this, statement, e => { this.showError(statement, e) })
            case "press_key":
                return await press_key.call(this, statement, e => { this.showError(statement, e) })
            case "wait_for":
                return await wait_for.call(this, statement, e => { this.showError(statement, e) })
            case "scroll":
                return await scroll.call(this, statement, e => { this.showError(statement, e) })

            
            // Logic
            case "controls_if":
                return await controls_if.call(this, statement, e => { this.showError(statement, e) })
            case "logic_compare":
                return await logic_compare.call(this, statement, e => { this.showError(statement, e) })
            case "logic_operation":
                return await logic_operation.call(this, statement, e => { this.showError(statement, e) })
            case "logic_negate":
                return await logic_negate.call(this, statement, e => { this.showError(statement, e) })
            case "logic_boolean":
                return await logic_boolean.call(this, statement, e => { this.showError(statement, e) })
            case "logic_null":
                return null
            case "logic_ternary":
                return await logic_ternary.call(this, statement, e => { this.showError(statement, e) })

            
            // Loops
            case "controls_repeat_ext":
                return await controls_repeat_ext.call(this, statement, e => { this.showError(statement, e) })
            case "controls_whileUntil":
                return await controls_whileUntil.call(this, statement, e => { this.showError(statement, e) })
            case "controls_for":
                return await controls_for.call(this, statement, e => { this.showError(statement, e) })
            case "controls_forEach":
                return await controls_forEach.call(this, statement, e => { this.showError(statement, e) })
            case "controls_flow_statements":
                return await controls_flow_statements.call(this, statement, e => { this.showError(statement, e) })


            // Math
            case "math_number":
                return await math_number.call(this, statement, e => { this.showError(statement, e) })
            case "math_arithmetic":
                return await math_arithmetic.call(this, statement, e => { this.showError(statement, e) })
            case "math_single":
                return await math_single.call(this, statement, e => { this.showError(statement, e) })
            case "math_trig":
                return await math_trig.call(this, statement, e => { this.showError(statement, e) })
            case "math_constant":
                return await math_constant.call(this, statement, e => { this.showError(statement, e) })
            case "math_number_property":
                return await math_number_property.call(this, statement, e => { this.showError(statement, e) })
            case "convert_text_to_number":
                return await convert_text_to_number.call(this, statement, e => { this.showError(statement, e) })
            case "math_round":
                return await math_round.call(this, statement, e => { this.showError(statement, e) })
            case "math_on_list":
                return await math_on_list.call(this, statement, e => { this.showError(statement, e) })
            case "math_modulo":
                return await math_modulo.call(this, statement, e => { this.showError(statement, e) })
            case "math_constrain":
                return await math_constrain.call(this, statement, e => { this.showError(statement, e) })
            case "math_random_int":
                return await math_random_int.call(this, statement, e => { this.showError(statement, e) })
            case "math_random_float":
                return await math_random_float.call(this, statement, e => { this.showError(statement, e) })


            // Text
            case "text":
                return await text.call(this, statement, e => { this.showError(statement, e) })
            case "text_join":
                return await text_join.call(this, statement, e => { this.showError(statement, e) })
            case "text_append":
                return await text_append.call(this, statement, e => { this.showError(statement, e) })
            case "text_length":
                return await text_length.call(this, statement, e => { this.showError(statement, e) })
            case "text_isEmpty":
                return await text_isEmpty.call(this, statement, e => { this.showError(statement, e) })
            case "convert_to_text":
                return await convert_to_text.call(this, statement, e => { this.showError(statement, e) })
            case "text_indexOf":
                return await text_indexOf.call(this, statement, e => { this.showError(statement, e) })
            case "text_charAt":
                return await text_charAt.call(this, statement, e => { this.showError(statement, e) })
            case "text_getSubstring":
                return await text_getSubstring.call(this, statement, e => { this.showError(statement, e) })
            case "text_changeCase":
                return await text_changeCase.call(this, statement, e => { this.showError(statement, e) })
            case "text_trim":
                return await text_trim.call(this, statement, e => { this.showError(statement, e) })
            case "text_print":
                return await text_print.call(this, statement, e => { this.showError(statement, e) })
            case "prompt":
                return await prompt.call(this, statement, e => { this.showError(statement, e) })

            
            // Lists
            case "lists_create_with":
                return await lists_create_with.call(this, statement, e => { this.showError(statement, e) })
            case "lists_chooser":
                return await lists_chooser.call(this, statement, e => { this.showError(statement, e) })
            case "lists_repeat":
                return await lists_repeat.call(this, statement, e => { this.showError(statement, e) })
            case "lists_length":
                return await lists_length.call(this, statement, e => { this.showError(statement, e) })
            case "lists_isEmpty":
                return await lists_isEmpty.call(this, statement, e => { this.showError(statement, e) })
            case "lists_indexOf":
                return await lists_indexOf.call(this, statement, e => { this.showError(statement, e) })
            case "lists_getIndex":
                return await lists_getIndex.call(this, statement, e => { this.showError(statement, e) })
            case "lists_setIndex":
                return await lists_setIndex.call(this, statement, e => { this.showError(statement, e) })
            case "lists_getSublist":
                return await lists_getSublist.call(this, statement, e => { this.showError(statement, e) })
            case "lists_split":
                return await lists_split.call(this, statement, e => { this.showError(statement, e) })
            case "lists_regex":
                return await lists_regex.call(this, statement, e => { this.showError(statement, e) })
            case "lists_sort":
                return await lists_sort.call(this, statement, e => { this.showError(statement, e) })


            // Variables
            case "variables_set":
                return await variables_set.call(this, statement, e => { this.showError(statement, e) })
            case "variables_get":
                return await variables_get.call(this, statement, e => { this.showError(statement, e) })
            case "math_change":
                return await math_change.call(this, statement, e => { this.showError(statement, e) })


            default:
                return null
        }
    }
}


export { Piper }
